presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java

```diff
@@ -697,36 +706,67 @@ private void queryFinished(ManagedQueryExecution query)
                     group = group.parent.get();
                 }
             }
+
             updateEligibility();
         }
     }
 
-    // Memory usage stats are expensive to maintain, so this method must be called periodically to update them
-    protected void internalRefreshStats()
+    protected ResourceUsage updateResourceUsageAndGetDelta()
     {
         checkState(Thread.holdsLock(root), "Must hold lock to refresh stats");
         synchronized (root) {
+            ResourceUsage groupUsageDelta = new ResourceUsage(0L, 0L);
+
             if (subGroups.isEmpty()) {
-                cachedMemoryUsageBytes = 0;
-                for (ManagedQueryExecution query : runningQueries) {
-                    cachedMemoryUsageBytes += query.getUserMemoryReservation().toBytes();
+                // Leaf resource group
+                for (ManagedQueryExecution query : runningQueries.keySet()) {
+                    ResourceUsage oldResourceUsage = runningQueries.get(query);
+                    ResourceUsage newResourceUsage = new ResourceUsage(
+                            query.getTotalCpuTime().toMillis(),
+                            query.getTotalMemoryReservation().toBytes());
+
+                    // Compute delta and update usage
+                    ResourceUsage queryUsageDelta = newResourceUsage.subtract(oldResourceUsage);
+                    runningQueries.put(query, newResourceUsage);
+                    groupUsageDelta = groupUsageDelta.add(queryUsageDelta);
                 }
             }
             else {
+                // Intermediate resource group
                 for (Iterator<InternalResourceGroup> iterator = dirtySubGroups.iterator(); iterator.hasNext(); ) {
                     InternalResourceGroup subGroup = iterator.next();
-                    long oldMemoryUsageBytes = subGroup.cachedMemoryUsageBytes;
-                    cachedMemoryUsageBytes -= oldMemoryUsageBytes;
-                    subGroup.internalRefreshStats();
-                    cachedMemoryUsageBytes += subGroup.cachedMemoryUsageBytes;
-                    if (!subGroup.isDirty()) {
-                        iterator.remove();
-                    }
-                    if (oldMemoryUsageBytes != subGroup.cachedMemoryUsageBytes) {
-                        subGroup.updateEligibility();
-                    }
+
+                    ResourceUsage subGroupUsageDelta = subGroup.updateResourceUsageAndGetDelta();
+                    groupUsageDelta = groupUsageDelta.add(subGroupUsageDelta);
                 }
             }
+
+            cachedResourceUsage = cachedResourceUsage.add(groupUsageDelta);
+            return groupUsageDelta;
+        }
+    }
+
+    /**
+     * Traverse all nodes in the subtree rooted at this node and update their eligibility. The traversal happens in
+     * a post-order fashion, i.e. eligibility of all the children is updated before the node itself.
+     */
+    protected void updateEligibilityAll()
```

Comment:
This is a very confusing method name. Perhaps we could rename both this and the existing `updateEligibility` to something that reflects which direction the recursion is going (upward/downward) or (parent/children).

Reply:
Updating method names to `updateEligibilityOnPathToRoot` and `updateEligibilityInSubtree`.